rclrs/src/node.rs

```diff
@@ -360,6 +360,84 @@ impl Node {
         domain_id
     }
 
+    /// Creates a `ParameterBuilder` that can be used to set parameter declaration options and
+    /// declare a parameter as `Optional`, `Mandatory` or `ReadOnly`.
+    ///
+    /// # Example
+    /// ```
+    /// # use rclrs::{Context, ParameterRange, RclrsError};
+    /// let context = Context::new([])?;
+    /// let node = rclrs::create_node(&context, "domain_id_node")?;
+    /// // Set it to a range of 0-100, with a step of 2
+    /// let range = ParameterRange {
+    ///     lower: Some(-100),
+    ///     upper: Some(100),
+    ///     step: Some(2),
+    /// };
+    /// let param = node.declare_parameter("int_param", 10)
+    ///                 .range(range)
+    ///                 .mandatory()
+    ///                 .unwrap();
+    /// assert_eq!(param.get(), 10);
+    /// param.set(50).unwrap();
+    /// assert_eq!(param.get(), 50);
+    /// // Out of range, will return an error
+    /// assert!(param.set(200).is_err());
+    /// # Ok::<(), RclrsError>(())
+    /// ```
+    pub fn declare_parameter<T: Declarable>(
+        &self,
+        name: &str,
+        default_value: T,
+    ) -> ParameterBuilder<'_, T> {
+        self._parameter.declare(name, default_value)
+    }
+
+    /// Creates a `ParameterBuilder` that can be used to set parameter declaration options and
+    /// declare a parameter from an iterable as `Optional`, `Mandatory` or `ReadOnly`.
+    pub fn declare_parameter_from_iter<U: IntoIterator>(
+        &self,
+        name: &str,
+        default_value: U,
+    ) -> ParameterBuilder<'_, Arc<[U::Item]>>
+    where
+        Arc<[U::Item]>: ParameterVariant,
+    {
+        self._parameter.declare_from_iter(name, default_value)
+    }
+
+    /// Creates a `ParameterBuilder` that can be used to set parameter declaration options and
+    /// declare a string array parameter as `Optional`, `Mandatory` or `ReadOnly`.
+    pub fn declare_string_array_parameter<U>(
+        &self,
+        name: &str,
+        default_value: U,
+    ) -> ParameterBuilder<'_, Arc<[Arc<str>]>>
+    where
+        U: IntoIterator,
+        U::Item: Into<Arc<str>>,
+    {
+        self._parameter.declare_string_array(name, default_value)
+    }
+
+    /// Helper function to declare an `Optional` parameter with a default unset value.
+    pub fn declare_unset_parameter<T: ParameterVariant>(
+        &self,
+        name: &str,
+    ) -> ParameterBuilder<'_, Option<T>> {
+        self._parameter.declare::<Option<T>>(name, None)
+    }
+
+    /// Enables usage of undeclared parameters for this node.
+    ///
+    /// Returns a `Parameter` struct that can be used to get and set all parameters.
+    pub fn use_undeclared_parameters(&self) -> Parameters {
+        self._parameter.allow_undeclared();
```

Comment:
In the current API, once undeclared parameters are allowed, they're allowed for the rest of the node's lifespan. It's hard for me to imagine a use case for undoing that allowance or even what the behavior should be to undo it.

In rclcpp this decision is set in stone at initialization, so I think it's reasonable for us to not offer a way to reverse it.

Reply:
But I don't see the `allow_undeclared` atomic being read anywhere.

Reply:
It has been added in anticipation of parameter services. When parameter services are implemented on top of this in the future, this flag will determine whether to reject parameter service requests on undeclared parameters.

Until then it's true that it doesn't have any effect, simply because we've designed the rclrs API in a way that makes it unnecessary to check this for rclrs API calls.